src/group_theory/group_ring.lean

```diff
@@ -31,130 +30,198 @@ group ring, group cohomology, monoid algebra
+  {M : Type*} [add_comm_monoid M] [has_scalar G M] [module k M] :
+  is_scalar_tower k (group_ring k G) M :=
+{ smul_assoc := λ g x m, show finsupp.total _ _ _ _ _ = g • finsupp.total _ _ _ _ _,
+    by simp only [ring_hom.id_apply, linear_map.map_smulₛₗ] }
+
+/-- If `M` has compatible `G`-module and `k`-module structures for a semiring `k`,
+it is also a `k[G]`-module. -/
+def action_to_module {k : Type*} [semiring k] {G : Type*} [monoid G]
+  {M : Type*} [add_comm_monoid M] [distrib_mul_action G M] [module k M]
+  [smul_comm_class k G M] : module (group_ring k G) M :=
+{ smul := (•),
+  one_smul := λ b, by
+  { dsimp,
+    show finsupp.total _ _ _ _ _ = _,
+    erw [finsupp.total_single],
+    simp only [one_smul] },
+  mul_smul := λ g h m, by
+  { refine g.induction_on _ _ _,
+    { intros g,
+      dsimp,
+      rw [monoid_algebra.mul_def, finsupp.sum_single_index],
+      { simp only [smul_def, one_mul, one_smul, finsupp.total_single],
+        simp only [finsupp.total_single, linear_map.map_finsupp_sum],
+        rw [finsupp.total_apply, finsupp.smul_sum],
+        simp only [mul_smul, smul_comm] },
+      { simp }},
+    { intros x y hx hy,
+      simp only [smul_def, add_mul, linear_map.map_add] at ⊢ hx hy,
+      simp only [hx, hy] },
+    { intros r f hf,
+      rw [smul_assoc, smul_mul_assoc, smul_assoc, hf] }},
+  smul_add := λ g b c, by
+    { simp only [smul_def, smul_add, finsupp.total_apply, finsupp.sum_add], },
+  smul_zero := λ x, by
+    simp only [smul_def, finsupp.total_apply, finsupp.sum_zero, smul_zero],
+  add_smul := λ r s x, linear_map.map_add _ _ _,
+  zero_smul := λ x, linear_map.map_zero _ }
 
-@[simp] lemma of_apply (g : G) :
-  of G g = finsupp.single g (1 : ℤ) := rfl
+end
 
-lemma zsmul_single_one (g : G) (r : ℤ) :
-  r • (of G g) = finsupp.single g r :=
-by simp only [mul_one, finsupp.smul_single', zsmul_eq_smul, of_apply]
+section
 
-@[elab_as_eliminator]
-lemma induction_on {p : group_ring G → Prop} (f : group_ring G)
-  (hM : ∀ g, p (of G g)) (hadd : ∀ f g : group_ring G, p f → p g → p (f + g))
-  (hsmul : ∀ (r : ℤ) f, p f → p (r • f)) : p f :=
-monoid_algebra.induction_on _ hM hadd hsmul
+variables {k : Type*} [semiring k] {G : Type*} [monoid G]
 
-lemma ext {P : Type*} [add_comm_group P] (f : group_ring G →+ P)
-  (g : group_ring G →+ P) (H : ∀ x, f (of G x) = g (of G x)) {x} :
+lemma smul_of_eq_single (g : G) (r : k) :
+  r • (of k G g) = finsupp.single g r :=
+by simp only [mul_one, finsupp.smul_single', of_apply]
+
+lemma ext {P : Type*} [add_comm_monoid P] [module k P]
+  (f : group_ring k G →ₗ[k] P) (g : group_ring k G →ₗ[k] P)
+  (H : ∀ x, f (of k G x) = g (of k G x)) {x} :
   f x = g x :=
 begin
-  congr,
-  refine finsupp.add_hom_ext (λ y n, _),
-  simp only [←zsmul_single_one, add_monoid_hom.map_zsmul, H],
+  refine x.induction_on H _ _,
+  { intros y z hy hz,
+    rw [f.map_add, g.map_add, hy, hz] },
+  { intros r y hy,
+    rw [f.map_smul, g.map_smul, hy] }
 end
```

Comment:
I think this belongs in `monoid_algebra/basic.lean`, as:
```lean
@[ext]
lemma monoid_algebra.lhom_ext {M : Type*} [add_comm_monoid M] [module k M]
  (f : monoid_algebra k G →ₗ[k] M) (g : monoid_algebra k G →ₗ[k] M)
  (H : ∀ x, f (of k G x) = g (of k G x)) :
  f = g :=
begin
  ext,
  apply H,
end
```
Note that `ext` essentially already solve this, but this cleans up the `lsingle` mess it leaves behind.

Reply:
Actually, on second thoughts `of` is not simp-normal anyway, so really we probably don't need this lemma at all.